### PR TITLE
Updated-Lagrangian axisymmetric U-P elements with updated permeability in GeoMechanicsApp

### PR DESCRIPTION
**Pull request by Vahid-Galavi**
**📝 Description**

- Added Updated-Lagrangian axisymmetric U-P elements
- Permeability can be updated as a function of volumetric strain
- Some minor improvements and styling

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 1]
```diff
@@ -1244,6 +1235,33 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
...
+void UPwSmallStrainElement<TDim,TNumNodes>::
+    CalculatePermeabilityUpdateFactor( ElementVariables &rVariables)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-SmallStrainUPwDiffOrderElement::CalculatePermeabilityUpdateFactor") << std::endl;
+
+    const PropertiesType& rProp = this->GetProperties();
+
+    if (rProp.Has(PERMEABILITY_CHANGE_INVERSE_FACTOR) &&
+        rProp[PERMEABILITY_CHANGE_INVERSE_FACTOR] > 0.0) {
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp]:
```suggestion
    if (rProp[PERMEABILITY_CHANGE_INVERSE_FACTOR] > 0.0) {
```

If `PERMEABILITY_CHANGE_INVERSE_FACTOR` is not defined, 0.0 will be returned

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 2]
```diff
@@ -1244,6 +1235,33 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
     KRATOS_CATCH( "" )
 }
 
+//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwSmallStrainElement<TDim,TNumNodes>::
+    CalculatePermeabilityUpdateFactor( ElementVariables &rVariables)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-SmallStrainUPwDiffOrderElement::CalculatePermeabilityUpdateFactor") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 3]
```diff
@@ -1244,6 +1235,33 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
...
+        const double epsV = EquivalentStress.CalculateTrace(rVariables.StrainVector);
+        const double ePrevious = rProp[POROSITY] / (1.0 - rProp[POROSITY]);
+        const double eCurrent = (1.0 + ePrevious) * std::exp(epsV) - 1.0;
+        const double permLog10 = (eCurrent - ePrevious) * InverseCK;
+        rVariables.PermeabilityUpdateFactor = pow(10.0, permLog10);
+    } else {
+        rVariables.PermeabilityUpdateFactor = 1.0;
+    }
+
+    // KRATOS_INFO("1-SmallStrainUPwDiffOrderElement::CalculatePermeabilityUpdateFactor") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp [hunk 4]
```diff
@@ -1844,6 +1853,32 @@ void SmallStrainUPwDiffOrderElement::
...
+void SmallStrainUPwDiffOrderElement::
+    CalculatePermeabilityUpdateFactor( ElementVariables &rVariables)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-SmallStrainUPwDiffOrderElement::CalculatePermeabilityUpdateFactor") << std::endl;
+
+    const PropertiesType& rProp = this->GetProperties();
+
+    if (rProp.Has(PERMEABILITY_CHANGE_INVERSE_FACTOR) &&
+        rProp[PERMEABILITY_CHANGE_INVERSE_FACTOR] > 0.0) {
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp]:
```suggestion
    if (rProp[PERMEABILITY_CHANGE_INVERSE_FACTOR] > 0.0) {
```

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp [hunk 5]
```diff
@@ -1844,6 +1853,32 @@ void SmallStrainUPwDiffOrderElement::
     KRATOS_CATCH( "" )
 }
 
+//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+void SmallStrainUPwDiffOrderElement::
+    CalculatePermeabilityUpdateFactor( ElementVariables &rVariables)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-SmallStrainUPwDiffOrderElement::CalculatePermeabilityUpdateFactor") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp [hunk 6]
```diff
@@ -1844,6 +1853,32 @@ void SmallStrainUPwDiffOrderElement::
...
+        const double epsV = EquivalentStress.CalculateTrace(rVariables.StrainVector);
+        const double ePrevious = rProp[POROSITY] / (1.0 - rProp[POROSITY]);
+        const double eCurrent = (1.0 + ePrevious) * std::exp(epsV) - 1.0;
+        const double permLog10 = (eCurrent - ePrevious) * InverseCK;
+        rVariables.PermeabilityUpdateFactor = pow(10.0, permLog10);
+    } else {
+        rVariables.PermeabilityUpdateFactor = 1.0;
+    }
+
+    // KRATOS_INFO("1-SmallStrainUPwDiffOrderElement::CalculatePermeabilityUpdateFactor") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:1876]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_FIC_element.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+//----------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwUpdatedLagrangianAxisymmetricFICElement<TDim,TNumNodes>::
+    CalculateBMatrix(Matrix& rB,
+                     const Matrix& GradNpT,
+                     const Vector& Np)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-UPwUpdatedLagrangianAxisymmetricFICElement::CalculateBMatrix()") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_FIC_element.cpp:49]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_FIC_element.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        const IndexType index = TDim * i;
+
+        rB( INDEX_2D_PLANE_STRAIN_XX, index + INDEX_X ) = GradNpT( i, INDEX_X );
+        rB( INDEX_2D_PLANE_STRAIN_YY, index + INDEX_Y ) = GradNpT( i, INDEX_Y );
+        rB( INDEX_2D_PLANE_STRAIN_ZZ, index + INDEX_X ) = Np[i]/radius;
+        rB( INDEX_2D_PLANE_STRAIN_XY, index + INDEX_X ) = GradNpT( i, INDEX_Y );
+        rB( INDEX_2D_PLANE_STRAIN_XY, index + INDEX_Y ) = GradNpT( i, INDEX_X );
+    }
+
+    // KRATOS_INFO("1-UPwUpdatedLagrangianAxisymmetricFICElement::CalculateBMatrix()") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_FIC_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+//----------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwUpdatedLagrangianAxisymmetricElement<TDim,TNumNodes>::
+    CalculateBMatrix(Matrix& rB,
+                     const Matrix& GradNpT,
+                     const Vector& Np)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-UPwUpdatedLagrangianAxisymmetricElement::CalculateBMatrix()") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        const IndexType index = TDim * i;
+
+        rB( INDEX_2D_PLANE_STRAIN_XX, index + INDEX_X ) = GradNpT( i, INDEX_X );
+        rB( INDEX_2D_PLANE_STRAIN_YY, index + INDEX_Y ) = GradNpT( i, INDEX_Y );
+        rB( INDEX_2D_PLANE_STRAIN_ZZ, index + INDEX_X ) = Np[i]/radius;
+        rB( INDEX_2D_PLANE_STRAIN_XY, index + INDEX_X ) = GradNpT( i, INDEX_Y );
+        rB( INDEX_2D_PLANE_STRAIN_XY, index + INDEX_Y ) = GradNpT( i, INDEX_X );
+    }
+
+    // KRATOS_INFO("1-UPwUpdatedLagrangianAxisymmetricElement::CalculateBMatrix()") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_axisymmetric_element.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+//----------------------------------------------------------------------------------------
+void UpdatedLagrangianUPwDiffOrderAxisymmetricElement::
+    CalculateBMatrix(Matrix& rB,
+                     const Matrix& GradNpT,
+                     const Vector& Np)
+{
+    KRATOS_TRY
+    // KRATOS_INFO("0-UpdatedLagrangianUPwDiffOrderAxisymmetricElement::CalculateBMatrix()") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_axisymmetric_element.cpp]:
```suggestion
```

File: applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_axisymmetric_element.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        const IndexType index = Dim * i;
+
+        rB( INDEX_2D_PLANE_STRAIN_XX, index + INDEX_X ) = GradNpT( i, INDEX_X );
+        rB( INDEX_2D_PLANE_STRAIN_YY, index + INDEX_Y ) = GradNpT( i, INDEX_Y );
+        rB( INDEX_2D_PLANE_STRAIN_ZZ, index + INDEX_X ) = Np[i]/radius;
+        rB( INDEX_2D_PLANE_STRAIN_XY, index + INDEX_X ) = GradNpT( i, INDEX_Y );
+        rB( INDEX_2D_PLANE_STRAIN_XY, index + INDEX_Y ) = GradNpT( i, INDEX_X );
+    }
+
+    // KRATOS_INFO("1-UpdatedLagrangianUPwDiffOrderAxisymmetricElement::CalculateBMatrix()") << std::endl;
```
Comment (miguelmaso) [inline applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_axisymmetric_element.cpp]:
```suggestion
```

**Review by miguelmaso — COMMENTED** [review]
Just some minor comments

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp (hunk 1 above)
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp]: I thought we receive a segmentation fault. Then, I can remove it

**Review by miguelmaso — APPROVED** [review]
I assume tests are passing.
It would be a good practice to add a patch test for the new elements

**Review by loumalouomega — APPROVED** [review]

**Comment by Vahid-Galavi** [comment]
Thanks @miguelmaso and @loumalouomega for the reviewing.

**Pull request merged**